plugins/inputs/azure_monitor/azure_monitor.go

```diff
@@ -0,0 +1,856 @@
+		return fmt.Errorf("client_secret is empty or missing. Please check your configuration")
+	}
+
+	if am.TenantID == "" {
+		return fmt.Errorf("tenant_id is empty or missing. Please check your configuration")
+	}
+
+	if len(am.ResourceTargets) == 0 && len(am.ResourceGroupTargets) == 0 && len(am.SubscriptionTargets) == 0 {
+		return fmt.Errorf("there is no target to collect metrics from. Please check your configuration")
+	}
+
+	for index, resourceTarget := range am.ResourceTargets {
+		if resourceTarget.ResourceID == "" {
+			return fmt.Errorf(
+				"resource target #%d resource_id is empty or missing. Please check your configuration", index+1)
+		}
+	}
+
+	for resourceGroupIndex, resourceGroupTarget := range am.ResourceGroupTargets {
+		if resourceGroupTarget.ResourceGroup == "" {
+			return fmt.Errorf(
+				"resource group target #%d resource_group is empty or missing. Please check your configuration",
+				resourceGroupIndex+1)
+		}
+
+		for resourceIndex, resource := range resourceGroupTarget.Resources {
+			if resource.ResourceType == "" {
+				return fmt.Errorf(
+					"resource group target #%d resource #%d resource_type is empty or missing. Please check your configuration",
+					resourceGroupIndex+1, resourceIndex+1)
+			}
+		}
+	}
+
+	for index, target := range am.SubscriptionTargets {
+		if target.ResourceType == "" {
+			return fmt.Errorf(
+				"subscription target #%d resource_type is empty or missing. Please check your configuration", index+1)
+		}
+	}
+
+	return nil
+}
+
+func (am *AzureMonitor) createResourceGroupTargetsFromSubscriptionTargets() error {
+	if len(am.SubscriptionTargets) == 0 {
+		return nil
+	}
+
+	am.Log.Info("Creating resource group targets from subscription targets")
+
+	apiURL := am.buildSubscriptionResourceGroupsAPIURL()
+	body, err := am.getTargetResponseBody(apiURL)
+
+	if err != nil {
+		return err
+	}
+
+	bodyData, err := unmarshalJSON(body)
+
+	if err != nil {
+		return err
+	}
+
+	for _, value := range bodyData["value"].([]interface{}) {
+		resourceGroup := value.(map[string]interface{})["name"].(string)
+		resourceGroupTarget := NewResourceGroupTarget(resourceGroup, am.SubscriptionTargets)
+
+		am.ResourceGroupTargets = append(am.ResourceGroupTargets, resourceGroupTarget)
+	}
+
+	return nil
+}
+
+func (am *AzureMonitor) createResourceTargetsFromResourceGroupTargets() error {
+	var waitGroup sync.WaitGroup
+
+	errChan := make(chan error, 1)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	defer cancel()
+
+	for index, target := range am.ResourceGroupTargets {
+		am.Log.Info("Creating resource targets from resource group target #", index+1)
+		waitGroup.Add(1)
+
+		go func(target *ResourceGroupTarget) {
+			defer waitGroup.Done()
+
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			err := am.createResourceTargetFromResourceGroupTarget(target)
+
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+
+				cancel()
+				return
+			}
+		}(target)
```

Comment:
Is it really necessary to do this in parallel? Remember we are still in `Init()` i.e. in the startup-phase of telegraf... If it's not absolutely necessary to do it in parallel, code can be simplified especially the error handling.